File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/ServiceAccessViewModel.kt
// Admin orchestrator for per-service access gating (docs/service-access-gating.md
// §v2 hardening). Mirror of iOS ServiceAccessViewModel + the webapp
// views/service-access.js:
//   - read the TRUE mode from the box, toggle open <-> restricted (owner-IRK),
//   - allow-list: ADD a person across THREE tiers (personal auto-approve /
//     personal manual-approve / group multi-use) -> seal {name,photo?} under the
//     household key -> AID-sign the create -> POST .com -> return the share-link
//     (+ author AID embedded so the friend derives a per-author contact AID),
//   - LIST (owner-signed; decrypt bundle locally; a group is ONE "label — k/N"
//     entry), REMOVE (AID-signed .com revoke + owner-IRK box allow-prune; group =
//     prune the whole inviteId; per-member = prune one bound AID),
//   - FINALIZE a manual-approve acceptance the friend replied back.
//
// v2 signing: create / revoke / the list query move from the IRK to the STABLE
// AID (the box-as-authority verifies the create against the owner AID, which —
// unlike the rotatable IRK — survives device rotation). set-mode + allow-remove
// stay owner-IRK (the box verifies its config-pinned owner IRK on those).

package com.flagshipserver.app.viewmodels

import android.content.Context
import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.ServiceAccessClient
import com.flagshipserver.app.api.ServiceInviteRow
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.core.InviteLink
import com.flagshipserver.app.core.ServiceInvite
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

sealed interface ServiceAccessPhase {
    data object Idle : ServiceAccessPhase
    data object Loading : ServiceAccessPhase
    data object Ready : ServiceAccessPhase
    data class Failed(val message: String) : ServiceAccessPhase
}

/** The three invite tiers, chosen at create time (docs §v2 Phase 3). */
enum class InviteTier {
    /** First-bind, auto-approve — the casual fast path (default). */
    PERSONAL_AUTO,
    /** Sensitive: redeem → pending → friend replies an acceptance → author
     *  finalizes. Closes the link-theft race without disclosing the friend. */
    PERSONAL_MANUAL,
    /** One link, up to maxRedemptions (0 = unlimited), auto-approve, lower-trust. */
    GROUP,
}

data class AccessPerson(
    val inviteId: String,
    val name: String,
    val photo: String?,
    val bound: Boolean,
    /** The friend's bound AID (lowercase hex) once they've redeemed; null while
     *  the invite is unredeemed. Removing a BOUND person also prunes this AID
     *  from the box's allow-list. */
    val boundAidHex: String? = null,
    /** GROUP tier: this row stands for the whole multi-use invite. */
    val isGroup: Boolean = false,
    /** GROUP: live redemptions / cap (0 = unlimited) — rendered "k/N". */
    val redemptions: Int = 0,
    val maxRedemptions: Int = 0,
    /** GROUP: the bound member AIDs (for per-member removal). */
    val memberAids: List<String> = emptyList(),
)

class ServiceAccessViewModel(
    private val serverDomain: String,
    private val serviceRef: String,
    private val username: String,
    private val client: ServiceAccessClient = ServiceAccessClient(),
    /** Owner-IRK signer (set-mode + the box allow-remove — box config-pinned IRK). */
    // Slice D (D-2) — changing WHO may access a service (set-mode + allow-remove)
    // is SENSITIVE membership mutation (membership.ts / serviceInvites.ts gate it
    // on the admin master root): sign with the admin root when held, else the
    // owner IRK (legacy). Canonical bytes unchanged. (The AID-signed create/
    // revoke path below is a separate identity and stays on the AID.)
    private val irkSigner: suspend (reason: String) -> Ed25519Sign = { r -> Keystore.adminSigningKey(r) },
    /** Stable-AID signer (create + revoke + the owner-signed .com list — v2). */
    private val aidSigner: suspend (reason: String) -> Ed25519Sign = { r -> Keystore.deriveAccountId(r) },
    /** Stable AID pub (hex) — the author identity recorded in invites. */
    private val aidPubHex: suspend (reason: String) -> String = { r -> Keystore.accountIdPubHex(r) },
    /** Household key — seals/opens the {name, photo?} bundle. */
    private val householdKey: suspend (reason: String) -> ByteArray = { r -> Keystore.deriveHouseholdKey(r) },
    private val now: () -> Long = { System.currentTimeMillis() },
) : ViewModel() {
    private val _phase = MutableStateFlow<ServiceAccessPhase>(ServiceAccessPhase.Idle)
    val phase: StateFlow<ServiceAccessPhase> = _phase.asStateFlow()

    private val _restricted = MutableStateFlow(false)
    val restricted: StateFlow<Boolean> = _restricted.asStateFlow()

    private val _allowCount = MutableStateFlow(0)
    val allowCount: StateFlow<Int> = _allowCount.asStateFlow()

    private val _people = MutableStateFlow<List<AccessPerson>>(emptyList())
    val people: StateFlow<List<AccessPerson>> = _people.asStateFlow()

    private val _lastLink = MutableStateFlow<String?>(null)
    val lastInviteLink: StateFlow<String?> = _lastLink.asStateFlow()

    private val _busyMode = MutableStateFlow(false)
    val busyMode: StateFlow<Boolean> = _busyMode.asStateFlow()
    private val _busyAdd = MutableStateFlow(false)
    val busyAdd: StateFlow<Boolean> = _busyAdd.asStateFlow()
    private val _busyFinalize = MutableStateFlow(false)
    val busyFinalize: StateFlow<Boolean> = _busyFinalize.asStateFlow()

    suspend fun load() {
        _phase.value = ServiceAccessPhase.Loading
        try {
            val state = client.getAccessState(serverDomain, serviceRef)
            _restricted.value = state.isRestricted
            _allowCount.value = state.allowCount
            _phase.value = ServiceAccessPhase.Ready
            if (state.isRestricted) refreshPeople()
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Couldn't reach the box to load access settings.")
        }
    }

    /** Owner-IRK-sign + POST the mode change. Returns success. */
    suspend fun setMode(restricted: Boolean): Boolean {
        if (_busyMode.value) return false
        _busyMode.value = true
        val mode = if (restricted) "restricted" else "open"
        return try {
            val key = irkSigner(
                if (restricted) "Restrict $serviceRef to your allow-list"
                else "Open $serviceRef to anyone with the link",
            )
            val ts = now()
            val bytes = ServiceInvite.canonicalSetAccessMode(serverDomain, serviceRef, mode, ts)
            val sig = key.sign(bytes)
            val request: JsonObject = buildJsonObject {
                put("serverId", JsonPrimitive(serverDomain))
                put("serviceRef", JsonPrimitive(serviceRef))
                put("mode", JsonPrimitive(mode))
                put("issuedAt", JsonPrimitive(ts))
            }
            client.setAccessMode(serverDomain, request, HexUtil.encode(sig))
            _restricted.value = restricted
            if (restricted) refreshPeople() else _people.value = emptyList()
            true
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Couldn't change who can open this. Try again in a moment.")
            false
        } finally {
            _busyMode.value = false
        }
    }

    /**
     * Mint a capability invite of the chosen [tier]; returns the share-link (also
     * in lastInviteLink). For GROUP, [maxRedemptions] (0 = unlimited) and an
     * optional [expiresAt] are committed in the create's signed bytes. The create
     * is AID-SIGNED (v2). The author AID is embedded in the link so the friend can
     * derive a per-author contact AID.
     */
    suspend fun addPerson(
        name: String,
        photo: String?,
        tier: InviteTier = InviteTier.PERSONAL_AUTO,
        maxRedemptions: Int = 0,
        expiresAt: Long? = null,
    ): String? {
        if (_busyAdd.value) return null
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return null
        _busyAdd.value = true
        return try {
            val reason = "Create an invite for $serviceRef"
            val key = aidSigner(reason)
            val aid = HexUtil.decode(aidPubHex(reason))!!
            val hh = householdKey(reason)
            val secret = ServiceInvite.randomSecret()
            val secretHash = ServiceInvite.secretHash(secret)
            // v2: random 128-bit id (no device-fingerprint leak).
            val inviteId = ServiceInvite.randomInviteId()
            val encryptedBundle = ServiceInvite.sealBundle(ServiceInvite.Bundle(trimmed, photo), hh, inviteId)
            val ts = now()
            val groupCap = if (tier == InviteTier.GROUP) maxRedemptions else null
            val groupExp = if (tier == InviteTier.GROUP) expiresAt else null
            val approvalMode = if (tier == InviteTier.PERSONAL_MANUAL) "manual" else "auto"
            val bytes = ServiceInvite.canonicalCreate(
                inviteId, aid, serviceRef, secretHash, encryptedBundle, ts, groupCap, groupExp,
            )
            val sig = key.sign(bytes)
            val request: JsonObject = buildJsonObject {
                put("inviteId", JsonPrimitive(inviteId))
                put("authorAID", JsonPrimitive(HexUtil.encode(aid)))
                put("serviceRef", JsonPrimitive(serviceRef))
                put("secretHash", JsonPrimitive(secretHash))
                put("encryptedBundle", JsonPrimitive(encryptedBundle))
                put("issuedAt", JsonPrimitive(ts))
                put("approvalMode", JsonPrimitive(approvalMode))
                if (groupCap != null) put("maxRedemptions", JsonPrimitive(groupCap))
                if (groupExp != null) put("expiresAt", JsonPrimitive(groupExp))
            }
            client.createInvite(username, request, HexUtil.encode(sig))
            // Carry the inviteId (&i=) ONLY for a manual-approve invite — the
            // friend signs the acceptance over it. Auto/group links stay bare.
            val linkInviteId = if (tier == InviteTier.PERSONAL_MANUAL) inviteId else null
            val link = InviteLink.shareLink(serverDomain, HexUtil.encode(secret), HexUtil.encode(aid), linkInviteId)
            _lastLink.value = link
            refreshPeople()
            link
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Couldn't create the invite. Try again in a moment.")
            null
        } finally {
            _busyAdd.value = false
        }
    }

    /**
     * MANUAL-approve finalize (v2 Phase 3 tier 2). The author pastes/scans the
     * acceptance their friend replied through the private channel; we submit it —
     * the friend's AID-signed acceptance + the owner's signed create (both carried
     * IN the reply) — to the AUTHOR's OWN box, which verifies both then binds the
     * contact AID. Returns true on a bind. The consumer's username is never seen.
     */
    suspend fun finalizeAcceptance(rawReply: String): Boolean {
        if (_busyFinalize.value) return false
        val parsed = InviteLink.decodeAcceptance(rawReply) ?: run {
            _phase.value = ServiceAccessPhase.Failed("That acceptance reply is malformed. Ask your contact to send it again.")
            return false
        }
        _busyFinalize.value = true
        return try {
            // ONLY {accept, acceptSig} — the box fetches the owner's signed create
            // from .com by inviteId (any-device finalize; no local create cache).
            client.acceptInvite(
                serverDomain,
                accept = parsed.accept,
                acceptSigHex = parsed.acceptSigHex,
            )
            refreshPeople()
            true
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Couldn't confirm them. Make sure the reply is from your invite and try again.")
            false
        } finally {
            _busyFinalize.value = false
        }
    }

    /** Revoke a person's (or a whole group's) access. Always records the `.com`
     *  revoke (now AID-signed). When the person has REDEEMED (bound AID), ALSO
     *  owner-IRK-prunes that AID from the box's allow-list (`.com` does NOT push to
     *  the daemon, so the box call is what actually denies them). A GROUP revoke
     *  prunes EVERY bound member AID. Both legs run; a box-prune failure surfaces. */
    suspend fun remove(inviteId: String, boundAidHex: String? = null, memberAids: List<String> = emptyList()) {
        val ts = now()
        // 1) `.com` revoke — records the revocation (author-AID signed, v2).
        val comOk = try {
            val key = aidSigner("Remove this person from $serviceRef")
            val bytes = ServiceInvite.canonicalRevoke(inviteId, ts)
            val sig = key.sign(bytes)
            val request: JsonObject = buildJsonObject {
                put("inviteId", JsonPrimitive(inviteId))
                put("issuedAt", JsonPrimitive(ts))
            }
            client.revokeInvite(username, request, HexUtil.encode(sig))
            true
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Couldn't remove them. Try again in a moment.")
            false
        }

        // 2) Box allow-list prune — for every bound AID (group: all members; else
        //    the single bound AID). This is what actually reaches the box.
        val targets = (memberAids + listOfNotNull(boundAidHex))
            .mapNotNull { it.lowercase().takeIf { a -> a.isNotEmpty() } }
            .distinct()
        var boxOk = true
        if (targets.isNotEmpty()) {
            boxOk = pruneAids(serviceRef, targets, ts)
        }

        if (comOk && boxOk) refreshPeople()
    }

    /** Per-member removal from a GROUP (the group itself stays). Prunes ONE bound
     *  AID from the box (owner-IRK). `.com` keeps the inviteId (other members live). */
    suspend fun removeGroupMember(aidHex: String) {
        val aid = aidHex.lowercase()
        if (aid.isEmpty()) return
        if (pruneAids(serviceRef, listOf(aid), now())) refreshPeople()
    }

    /** Owner-IRK-prune a set of AIDs from the box allow-list. Returns true iff all
     *  pruned. One biometric signer call; each AID a separate signed envelope. */
    private suspend fun pruneAids(serviceRef: String, aids: List<String>, ts: Long): Boolean {
        return try {
            val key = irkSigner("Remove from $serviceRef")
            for (aid in aids) {
                val sig = ServiceInvite.signRemoveServiceAllow(serverDomain, serviceRef, aid, ts, key)
                val request: JsonObject = buildJsonObject {
                    put("serverId", JsonPrimitive(serverDomain))
                    put("serviceRef", JsonPrimitive(serviceRef))
                    put("aid", JsonPrimitive(aid))
                    put("issuedAt", JsonPrimitive(ts))
                }
                client.removeServiceAllow(serverDomain, request, HexUtil.encode(sig))
            }
            true
        } catch (e: Throwable) {
            _phase.value = ServiceAccessPhase.Failed("Removed them, but the server is still catching up. Try again in a moment.")
            false
        }
    }

    /** List the author's live invites for this service from .com (OWNER-SIGNED),
     *  decrypting each bundle locally with the household key. A GROUP invite folds
     *  to ONE entry showing "k/N" + its bound members. */
    suspend fun refreshPeople() {
        try {
            val reason = "Show who can open $serviceRef"
            val aidPub = aidPubHex(reason)
            val hh = householdKey(reason)
            val ts = now()
            val sig = ServiceInvite.signServiceInviteListQuery(username, aidPub.lowercase(), "list", 0, ts, aidSigner(reason))
            val rows = client.listInvites(username, aidPub, ts, HexUtil.encode(sig))
            _people.value = rows
                .filter { it.serviceRef == serviceRef && it.revokedAt == null }
                .map { row -> toPerson(row, hh) }
        } catch (e: Throwable) {
            // keep the last list; load-phase error already surfaced
        }
    }

    private fun toPerson(row: ServiceInviteRow, hh: ByteArray): AccessPerson {
        var name = "unknown"
        var photo: String? = null
        runCatching { ServiceInvite.openBundle(row.encryptedBundleHex, hh, row.inviteId) }
            .getOrNull()?.let { name = it.name; photo = it.photo }
        return if (row.isGroup) {
            AccessPerson(
                inviteId = row.inviteId,
                name = name,
                photo = photo,
                bound = row.boundAidsHex.isNotEmpty(),
                boundAidHex = null,
                isGroup = true,
                redemptions = if (row.redemptions > 0) row.redemptions else row.boundAidsHex.size,
                maxRedemptions = row.maxRedemptions ?: 0,
                memberAids = row.boundAidsHex,
            )
        } else {
            AccessPerson(
                inviteId = row.inviteId,
                name = name,
                photo = photo,
                bound = row.boundAidHex != null,
                boundAidHex = row.boundAidHex,
            )
        }
    }

    fun clearLink() { _lastLink.value = null }

    companion object {
        fun create(
            context: Context,
            serverDomain: String,
            serviceRef: String,
            username: String,
        ): ServiceAccessViewModel = ServiceAccessViewModel(serverDomain, serviceRef, username)
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/ServiceAccessViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// Admin orchestrator for per-service access gating (docs/service-access-gating.md,
/// "## v2 hardening").
///
/// On the server-detail "Who can open this" surface:
///   - reads the TRUE mode from the box (`GET /api/service-access/<ref>`),
///   - toggles open ⇄ restricted with an owner-IRK `set-service-access-mode`
///     envelope to the box's pinned pipe,
///   - manages the allow-list across the THREE invite tiers (personal auto /
///     personal manual / group multi-use): add a person/group (seal the bundle
///     under the household key → **AID-sign** the create → POST `.com` → return
///     the `https://<server>/invite#<secret>&a=<authorAID>[&i=…]` link),
///     list (an OWNER-SIGNED `.com` query; decrypt the bundle locally; group =
///     one "label — k/N" row), and remove (**AID-signed** revoke on `.com` + an
///     owner-IRK prune of each bound AID on the box).
///
/// v2 signing: the author signs create / revoke / the listing with the STABLE
/// **AID** (so the box-as-authority verifies against the owner key across IRK
/// rotations; `.com` dual-accepts AID|IRK). The box's mode toggle + allow-remove
/// stay owner-**IRK** signed (the box pins the owner IRK). The friend's recorded
/// identity is their PER-AUTHOR contact AID.
@Observable
@MainActor
public final class ServiceAccessViewModel {

    public enum Phase: Equatable, Sendable {
        case idle
        case loading
        case ready
        case failed(String)
    }

    /// The three invite tiers the create picker offers (docs §v2 Phase 3).
    public enum InviteTier: Equatable, Sendable {
        /// Personal, auto-approve — first-bind (the casual-leak window, default).
        case personalAuto
        /// Personal, manual-approve (sensitive) — the friend's redeem returns
        /// {pending}; the friend replies an AID-signed acceptance the AUTHOR
        /// finalizes. `approvalMode:"manual"` on `.com`.
        case personalManual
        /// Group / multi-use — one link, `maxRedemptions` (0 = unlimited) +
        /// optional `expiresAt`. Auto-approve, lower-trust by construction.
        case group(maxRedemptions: Int, expiresAt: Int64?)

        public var isGroup: Bool { if case .group = self { return true }; return false }
        public var isManual: Bool { self == .personalManual }
    }

    public struct Person: Equatable, Sendable, Identifiable {
        public let inviteId: String
        public let name: String
        public let photo: String?
        public let bound: Bool
        /// The friend's allow-listed AID (lowercase hex) once they've redeemed;
        /// nil for an unredeemed invite. Drives the box-side prune on remove.
        public let boundAID: String?
        /// GROUP rows: every bound AID (a group is a labeled set). The box prune
        /// removes ALL of them. Empty for a personal invite.
        public let groupBoundAIDs: [String]
        /// GROUP rows: the cap (0 = unlimited) so the row can show "k/N".
        public let groupMax: Int?
        /// Manual-approve invite (no bind until the author finalizes).
        public let manual: Bool
        public var id: String { inviteId }
        public var isGroup: Bool { groupMax != nil }
        public init(inviteId: String, name: String, photo: String?, bound: Bool, boundAID: String?, groupBoundAIDs: [String] = [], groupMax: Int? = nil, manual: Bool = false) {
            self.inviteId = inviteId
            self.name = name
            self.photo = photo
            self.bound = bound
            self.boundAID = boundAID
            self.groupBoundAIDs = groupBoundAIDs
            self.groupMax = groupMax
            self.manual = manual
        }
    }

    public private(set) var phase: Phase = .idle
    public private(set) var restricted = false
    public private(set) var allowCount = 0
    public private(set) var people: [Person] = []
    /// Set after a successful create so the screen can surface the share-link.
    public private(set) var lastInviteLink: String?
    /// The tier the last-created link was for (so the screen can label it).
    public private(set) var lastInviteTier: InviteTier?

    public var busyMode = false
    public var busyAdd = false

    private let client: any ServiceAccessClient
    private let serverDomain: String
    private let serviceRef: String
    private let username: String
    private let controlBase: URL
    /// One-biometric author v2 bundle: (AID signer, household key). The AID signs
    /// create/revoke/list AND is the recorded inviter identity; the household
    /// decrypts the listed bundles.
    private let authorAidKeys: @MainActor (String) async throws -> (aid: Curve25519.Signing.PrivateKey, household: Data)
    /// Owner IRK signer — the box mode toggle + allow-remove (the box pins the IRK).
    private let irkSigner: @MainActor (String) async throws -> Curve25519.Signing.PrivateKey
    private let now: () -> Int64

    public init(
        client: any ServiceAccessClient,
        serverDomain: String,
        serviceRef: String,
        username: String,
        controlBase: URL = Endpoints.controlBaseUrl,
        authorAidKeys: (@MainActor (String) async throws -> (aid: Curve25519.Signing.PrivateKey, household: Data))? = nil,
        irkSigner: (@MainActor (String) async throws -> Curve25519.Signing.PrivateKey)? = nil,
        now: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.client = client
        self.serverDomain = serverDomain
        self.serviceRef = serviceRef
        self.username = username
        self.controlBase = controlBase
        self.now = now
        self.authorAidKeys = authorAidKeys ?? { reason in
            try await Keystore.deriveInviteAuthorAidKeys(reason: reason)
        }
        self.irkSigner = irkSigner ?? { reason in try await Keystore.deriveIRK(reason: reason) }
    }

    public func load() async {
        phase = .loading
        do {
            let state = try await client.getAccessState(serverDomain: serverDomain, serviceRef: serviceRef)
            restricted = state.isRestricted
            allowCount = state.allowCount
            phase = .ready
            if restricted { await refreshPeople() }
        } catch {
            phase = .failed("Couldn't reach the box to load access settings.")
        }
    }

    /// Owner-IRK-sign + POST the mode change. Reverts `restricted` on failure.
    public func setMode(restricted want: Bool) async {
        if busyMode { return }
        busyMode = true
        defer { busyMode = false }
        let mode = want ? "restricted" : "open"
        do {
            let key = try await irkSigner(want
                ? "Restrict \(serviceRef) to your allow-list"
                : "Open \(serviceRef) to anyone with the link")
            let bytes = try ServiceInvite.canonicalSetAccessMode(
                serverId: serverDomain, serviceRef: serviceRef, mode: mode, issuedAt: now())
            let sig = try ServiceInvite.sign(bytes, with: key)
            let request: [String: Any] = ["serverId": serverDomain, "serviceRef": serviceRef, "mode": mode, "issuedAt": now()]
            _ = try await client.setAccessMode(serverDomain: serverDomain, request: request, signatureHex: HexUtil.encode(sig))
            restricted = want
            if want { await refreshPeople() } else { people = [] }
        } catch {
            phase = .failed("Couldn't change who can open this. Try again in a moment.")
        }
    }

    /// Mint a capability invite for a new person / group at the chosen tier;
    /// returns the share-link (also stored in `lastInviteLink`). One biometric
    /// (author AID + household). The link carries the author's AID (the friend
    /// derives their per-author contact AID) and — for a MANUAL invite — the
    /// inviteId (so the friend can sign the acceptance).
    @discardableResult
    public func addPerson(name: String, photo: String?, tier: InviteTier = .personalAuto) async -> String? {
        if busyAdd { return nil }
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        busyAdd = true
        defer { busyAdd = false }
        do {
            let keys = try await authorAidKeys("Create an invite for \(serviceRef)")
            let aidPub = keys.aid.publicKey.rawRepresentation
            let bundle = ServiceInvite.Bundle(name: trimmed, photo: photo)
            let secret = randomSecret()
            let secretHash = ServiceInvite.secretHash(secret: secret)
            // v2 §M2: a random 128-bit id (drops the devicePub fingerprint leak).
            let inviteId = ServiceInvite.randomServiceInviteId()
            let encryptedBundle = try ServiceInvite.sealBundle(bundle, householdKey: keys.household, inviteId: inviteId)
            let ts = now()

            var maxN: Int?
            var expiresAt: Int64?
            if case let .group(m, exp) = tier {
                maxN = m
                expiresAt = exp
            }
            let bytes = try ServiceInvite.canonicalCreate(
                inviteId: inviteId, authorAID: aidPub, serviceRef: serviceRef,
                secretHash: secretHash, encryptedBundle: encryptedBundle, issuedAt: ts,
                maxRedemptions: maxN, expiresAt: expiresAt)
            // Slice D (D-2) — service-invite CREATE is SENSITIVE (admin-gated).
            // Sign with the admin master root when this device holds one; else
            // AID (the legacy dual-accept path `.com`/box still honor when no
            // admin root is pinned). The `authorAID` field stays the AID (the
            // recorded inviter identity + the link's author key) — only the
            // signing key changes; canonical bytes are unchanged.
            let createSignKey: Curve25519.Signing.PrivateKey = Keystore.hasAdminRoot
                ? try await Keystore.adminRootKey(reason: "Create an invite for \(serviceRef)")
                : keys.aid
            let sig = try ServiceInvite.sign(bytes, with: createSignKey)
            var request: [String: Any] = [
                "inviteId": inviteId,
                "authorAID": HexUtil.encode(aidPub),
                "serviceRef": serviceRef,
                "secretHash": secretHash,
                "encryptedBundle": encryptedBundle,
                "issuedAt": ts,
            ]
            if let maxN { request["maxRedemptions"] = maxN }
            if let expiresAt { request["expiresAt"] = expiresAt }
            // approvalMode is a `.com`-side policy field (NOT in the signed bytes).
            if tier.isManual { request["approvalMode"] = "manual" }
            try await client.createInvite(controlBase: controlBase, username: username, request: request, signatureHex: HexUtil.encode(sig))

            // No local create cache: the author's box fetches the signed create
            // from `.com` at manual-finalize, so an invite can be finalized from
            // ANY of the author's devices.

            // The link carries the authorAID always; the inviteId only for manual
            // (the friend needs it to sign the out-of-band acceptance).
            let link = ServiceInviteLinks.inviteLink(
                serverDomain: serverDomain, secretHex: HexUtil.encode(secret),
                authorAidHex: HexUtil.encode(aidPub),
                inviteId: tier.isManual ? inviteId : nil)
            lastInviteLink = link
            lastInviteTier = tier
            await refreshPeople()
            return link
        } catch {
            phase = .failed("Couldn't create the invite. Try again in a moment.")
            return nil
        }
    }

    /// Remove a person / group: AID-signed revoke on `.com` (records the
    /// revocation) AND — for each bound AID — an owner-IRK prune from the box's
    /// allow-list. The box allow-list is add-only, so a `.com` revoke alone never
    /// reaches it: a redeemed friend (or every group member) would keep access
    /// without the prune. For a GROUP, EVERY bound AID is pruned in one op. A
    /// failure of any leg surfaces — losing a box prune silently would leave a
    /// friend with live access.
    public func remove(inviteId: String) async {
        let person = people.first(where: { $0.inviteId == inviteId })
        // AIDs to prune from the box: a group's whole set, else the single bind.
        var aids: [String] = person?.groupBoundAIDs ?? []
        if aids.isEmpty, let single = person?.boundAID, !single.isEmpty { aids = [single] }

        var revokeFailed = false
        var pruneFailed = false

        // `.com` revoke (AID-signed; records the revocation / what the directory shows).
        do {
            let aidKey = try await authorAidKeys("Remove this person from \(serviceRef)").aid
            let ts = now()
            let bytes = try ServiceInvite.canonicalRevoke(inviteId: inviteId, issuedAt: ts)
            // Slice D (D-2) — service-invite REVOKE is SENSITIVE (admin-gated):
            // admin master root when present, else the legacy AID.
            let revokeSignKey: Curve25519.Signing.PrivateKey = Keystore.hasAdminRoot
                ? try await Keystore.adminRootKey(reason: "Remove this person from \(serviceRef)")
                : aidKey
            let sig = try ServiceInvite.sign(bytes, with: revokeSignKey)
            let request: [String: Any] = ["inviteId": inviteId, "issuedAt": ts]
            try await client.revokeInvite(controlBase: controlBase, username: username, inviteId: inviteId, request: request, signatureHex: HexUtil.encode(sig))
        } catch {
            revokeFailed = true
        }

        // Box prune of each bound AID (owner-IRK signed — the box pins the IRK).
        if !aids.isEmpty {
            do {
                let key = try await irkSigner("Revoke access to \(serviceRef)")
                for aid in aids {
                    let ts = now()
                    let bytes = try ServiceInvite.canonicalRemoveServiceAllow(
                        serverId: serverDomain, serviceRef: serviceRef, aid: aid, issuedAt: ts)
                    let sig = try ServiceInvite.sign(bytes, with: key)
                    let request: [String: Any] = ["serverId": serverDomain, "serviceRef": serviceRef, "aid": aid.lowercased(), "issuedAt": ts]
                    _ = try await client.removeServiceAllow(serverDomain: serverDomain, request: request, signatureHex: HexUtil.encode(sig))
                }
            } catch {
                pruneFailed = true
            }
        }

        await refreshPeople()
        if pruneFailed {
            phase = .failed("Removed from the directory, but couldn't reach the box to revoke access. Try again so they're fully removed.")
        } else if revokeFailed {
            phase = .failed("Couldn't remove them. Try again in a moment.")
        }
    }

    /// List the author's live invites for this service from `.com` (OWNER-SIGNED
    /// query, v2 §C2), decrypting each bundle locally with the household key. A
    /// GROUP invite renders as ONE row carrying its whole bound set + cap.
    public func refreshPeople() async {
        do {
            let keys = try await authorAidKeys("Show who can open \(serviceRef)")
            let aidPubHex = HexUtil.encode(keys.aid.publicKey.rawRepresentation)
            let ts = now()
            // OWNER-SIGNED list query (scope "list", cursor 0).
            let qBytes = try ServiceInvite.canonicalListQuery(
                username: username, authorAID: aidPubHex, scope: "list", cursor: 0, issuedAt: ts)
            let qSig = try ServiceInvite.sign(qBytes, with: keys.aid)
            let query: [String: String] = [
                "authorAID": aidPubHex,
                "scope": "list",
                "cursor": "0",
                "issuedAt": String(ts),
            ]
            let rows = try await client.listInvites(
                controlBase: controlBase, username: username, query: query, signatureHex: HexUtil.encode(qSig))
            people = rows
                .filter { $0.serviceRef == serviceRef && $0.revokedAt == nil }
                .map { row in
                    var name = "unknown"
                    var photo: String?
                    if let b = try? ServiceInvite.openBundle(row.encryptedBundleHex, householdKey: keys.household, inviteId: row.inviteId) {
                        name = b.name
                        photo = b.photo
                    }
                    let groupAids = row.isGroup ? (row.boundAidsHex.isEmpty ? (row.boundAidHex.map { [$0] } ?? []) : row.boundAidsHex) : []
                    let bound = row.boundAidHex != nil || !row.boundAidsHex.isEmpty
                    return Person(
                        inviteId: row.inviteId,
                        name: name,
                        photo: photo,
                        bound: bound,
                        boundAID: row.boundAidHex,
                        groupBoundAIDs: groupAids,
                        groupMax: row.maxRedemptions,
                        manual: row.approvalMode == "manual")
                }
        } catch {
            // Keep the last-known list; the load-phase error is already surfaced.
        }
    }

    public func clearLink() { lastInviteLink = nil; lastInviteTier = nil }

    private func randomSecret() -> Data {
        var d = Data(count: 32)
        d.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        return d
    }
}
